pkg/volume/aws_ebs/aws_util.go

```diff
@@ -17,47 +17,58 @@ limitations under the License.
 package aws_ebs
 
 import (
-	"errors"
+	"fmt"
 	"os"
+	"path/filepath"
+	"strings"
 	"time"
 
 	"github.com/golang/glog"
-	aws_cloud "k8s.io/kubernetes/pkg/cloudprovider/providers/aws"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/aws"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/keymutex"
+	"k8s.io/kubernetes/pkg/util/sets"
 	"k8s.io/kubernetes/pkg/volume"
 )
 
+const (
+	diskPartitionSuffix = ""
+	diskXVDPath         = "/dev/xvd"
+	diskXVDPattern      = "/dev/xvd*"
+	maxChecks           = 60
+	maxRetries          = 10
+	checkSleepDuration  = time.Second
+	errorSleepDuration  = 5 * time.Second
+)
+
+// Singleton key mutex for keeping attach/detach operations for the same PD atomic
+var attachDetachMutex = keymutex.NewKeyMutex()
+
 type AWSDiskUtil struct{}
 
-// Attaches a disk specified by a volume.AWSElasticBlockStore to the current kubelet.
+// Attaches a disk to the current kubelet.
 // Mounts the disk to it's global path.
-func (util *AWSDiskUtil) AttachAndMountDisk(b *awsElasticBlockStoreBuilder, globalPDPath string) error {
-	volumes, err := b.getVolumeProvider()
+func (diskUtil *AWSDiskUtil) AttachAndMountDisk(b *awsElasticBlockStoreBuilder, globalPDPath string) error {
+	glog.V(5).Infof("AttachAndMountDisk(...) called for PD %q. Will block for existing operations, if any. (globalPDPath=%q)\r\n", b.volumeID, globalPDPath)
+
+	// Block execution until any pending detach operations for this PD have completed
+	attachDetachMutex.LockKey(b.volumeID)
+	defer attachDetachMutex.UnlockKey(b.volumeID)
+
+	glog.V(5).Infof("AttachAndMountDisk(...) called for PD %q. Awake and ready to execute. (globalPDPath=%q)\r\n", b.volumeID, globalPDPath)
+
+	xvdBefore, err := filepath.Glob(diskXVDPattern)
```

Comment:
All the `udevadm` stuff was a workaround for a GCE PD issue. It shouldn't hurt, but since it is unnecessary, we should remove it.